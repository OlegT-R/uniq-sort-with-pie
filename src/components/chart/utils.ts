interface IInitHtml {
  html: HTMLElement;
  context: CanvasRenderingContext2D | null;
}

// init base chart dom structure
export const initHtml = (): IInitHtml => {
  const html = document.createElement('div');
  html.id = 'piechart-block';

  const canvas = document.createElement('canvas');
  canvas.id = 'piechart';

  html.appendChild(canvas);

  const context = canvas.getContext('2d');
  return {
    html,
    context
  };
};

interface IDataSet {
  labels: string[];
  datasets: number[];
}

export const getChartDataSet = (chars: string): IDataSet => {
  const countMap: { [key: string]: number } = {};
  for (const ch of chars) {
    countMap[ch] = !countMap[ch] ? 1 : countMap[ch] + 1;
  }
  const labels = [];
  const datasets = [];
  for (const key in countMap) {
    if (countMap.hasOwnProperty(key)) {
      labels.push(key);
      datasets.push(countMap[key]);
    }
  }
  return {labels, datasets};
};

export const addData = (chart: any, labels: string[], data: number[]) => {
  chart.data.labels = labels;

  chart.data.datasets[0].backgroundColor = [];
  chart.data.datasets[0].data = data.map((value: number) => {
    chart.data.datasets[0].backgroundColor.push(
        // tslint:disable-next-line no-bitwise
        '#' + ((Math.random() * 0xffffff) << 0).toString(16)
    );
    return value;
  });
  chart.update();
};
