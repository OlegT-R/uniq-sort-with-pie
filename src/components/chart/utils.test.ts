import {getChartDataSet, initHtml} from './utils';

describe('chart utils test', () => {
  test('html structure', () => {
    const {html} = initHtml();
    const markup = html.outerHTML;
    expect(JSON.stringify({html: markup})).toMatchSnapshot();
  });

  test('getChartDataSet should return labels and dataset', () => {
    const {labels, datasets} = getChartDataSet('aaabbbcczk');
    expect(labels).toEqual(['a', 'b', 'c', 'z', 'k']);
    expect(datasets).toEqual([3, 3, 2, 1, 1]);

    const empty = getChartDataSet('');
    expect(empty.labels).toEqual([]);
    expect(empty.datasets).toEqual([]);
  });
});
