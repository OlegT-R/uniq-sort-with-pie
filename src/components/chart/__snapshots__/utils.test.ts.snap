// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chart utils test html structure 1`] = `"{\\"html\\":\\"<div id=\\\\\\"piechart-block\\\\\\"><canvas id=\\\\\\"piechart\\\\\\"></canvas></div>\\"}"`;
